concept/type_/constraint.rs

```diff
@@ -4,139 +4,548 @@
+    pub(crate) fn new(description: ConstraintDescription, source: T) -> Self {
+        Self { description, source }
+    }
+}
+
+impl<T: KindAPI<'static>> Constraint<T> for TypeConstraint<T> {
+    fn description(&self) -> ConstraintDescription {
+        self.description.clone()
+    }
+
+    fn source(&self) -> T {
+        self.source.clone()
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct CapabilityConstraint<CAP: Capability<'static>> {
+    description: ConstraintDescription,
+    source: CAP,
+}
+
+impl<CAP: Capability<'static>> CapabilityConstraint<CAP> {
+    pub(crate) fn new(description: ConstraintDescription, source: CAP) -> Self {
+        Self { description, source }
+    }
+}
+
+impl<CAP: Capability<'static>> Constraint<CAP> for CapabilityConstraint<CAP> {
+    fn description(&self) -> ConstraintDescription {
+        self.description.clone()
+    }
+
+    fn source(&self) -> CAP {
+        self.source.clone()
+    }
+}
+
+// Siblings = both interface types i1 and i2 are capabilities of the same capability type (owns/plays/relates)
+// of the same object type (e.g. they are owned by the same type, they are played by the same type)
+// with "i1 isa $x; i2 isa $x;"
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum ConstraintScope {
+    SingleInstanceOfType,
+    SingleInstanceOfTypeOrSubtype,
+    AllInstancesOfSiblingTypeOrSubtypes,
+    AllInstancesOfTypeOrSubtypes,
+}
+
+macro_rules! filter_by_constraint_category {
+    ($constraints_iter:expr, $constraint_category:ident) => {
+        $constraints_iter.filter(|constraint| {
+            constraint.category() == crate::type_::constraint::ConstraintCategory::$constraint_category
+        })
+    };
+}
+pub(crate) use filter_by_constraint_category;
+
+macro_rules! filter_by_source {
+    ($constraints_iter:expr, $source:expr) => {
+        $constraints_iter.filter(|constraint| &constraint.source() == &$source)
+    };
+}
+pub(crate) use filter_by_source;
+
+macro_rules! filter_by_scope {
+    ($constraints_iter:expr, $scope:expr) => {
+        $constraints_iter.filter(|constraint| constraint.scope() == $scope)
+    };
+}
+pub(crate) use filter_by_scope;
+
+macro_rules! filter_out_unchecked_constraints {
+    ($constraints_iter:expr) => {
+        $constraints_iter.filter(|constraint| !constraint.unchecked())
+    };
+}
+pub(crate) use filter_out_unchecked_constraints;
+
+use crate::type_::{owns::Owns, plays::Plays, relates::Relates, type_manager::TypeManager, Ordering};
+
+pub(crate) fn get_cardinality_constraints<C: Constraint<T>, T: Hash + Eq>(
+    constraints: impl IntoIterator<Item = C>,
+) -> HashSet<C> {
+    filter_by_constraint_category!(constraints.into_iter(), Cardinality).collect()
+}
+
+pub(crate) fn get_cardinality_constraint<'a, CAP: Capability<'static>>(
+    capability: CAP,
+    constraints: impl IntoIterator<Item = &'a CapabilityConstraint<CAP>>,
+) -> Option<CapabilityConstraint<CAP>> {
```

Comment:
That's the point of this API: you have a collection of constraints belonging to different capabilities, and you search for the constraint of a specific capability! For example, it's used to filter out inherited constraints, Can also be used for any other reason.

Capabilities themselves get their constraints using these methods. 
* The whole `constraint.rs` file is for defining `Constraint` and macros/functions to work with them, filter their collections, etc.
* `type_manager.rs` gets constraints from `type_reader.rs`, filters them using `constraint.rs`
* `owns.rs`/`plays.rs`/`relates.rs`/..../`entity_type.rs` call for `type_manager.rs` methods in their concept api public methods
* managers have data and schema validations, and sometimes these methods from `constraint.rs` are useful not only for concept api calls, but for complex algorithms of finding interesting constraints.

This API doesn't need to be tight and it doesn't even try to seem tight: it returns an `Option` because of the clear reason.

Reply:
I get your idea overall though and may have similar issues somewhere else, and maybe these methods are just badly formed so you misunderstood their purpose.